Improve StringTypeAdapter templates and handling

zdobersek opened the pull request:
#### bd28677dcc291ac6b56dfbd517849f993ea6f9fa
<pre>
Improve StringTypeAdapter templates and handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=250017">https://bugs.webkit.org/show_bug.cgi?id=250017</a>

Reviewed by NOBODY (OOPS!).

StringConcatenate&apos;s makeString() and makeAtomString() are adjusted to accept
arguments as const lvalue references, propagating them down to each
StringTypeAdapter object without unnecessary copies. Similarly, the adapter
objects too are handled through const lvalue references once constructed and
passed on.

Newly-imposed constness requires additional const qualifiers on common methods
across different StringTypeAdapter specializations.

The are8Bit() and stringTypeAdapterAccumulator() helper functions are simplified
through using parameter packs and fold expressions to perform the necessary
operations across a range of adapter objects.

StringTypeAdapter specialization for StringView is moved from the StringView
header into the StringConcatenate header. A new partial specialization for the
StringTypeAdapter template is added, conditionally enabled if a const reference
of the specified type can be used to construct a StringView object. This
specialization inherits from the StringView specialization, hence a StringView
is constructed for such objects.

An exception for this partial specialization are character string literals whose
type is reduced to a fixed-length C array. These for the moment require a
separate specialization in order to avoid implicitly creating a String object
that&apos;s then used for the StringView. The fixed-length array data is instead
passed to the partial specializations handling string pointers, in order to
pass the data through string-length computation and getting the correct length.
This avoids potential problems with fixed-length arrays with inflated size that
are used as buffers for string data that&apos;s otherwise null-terminated at some
previous position.

This enables removing a bunch of other StringTypeAdapter specializations.
Different WTF string and impl types can be used to construct the StringView,
as well as ASCIILiteral. For Vectors of character values, a corresponding Span
can be implicitly constructed and is then used for StringView construction.

In StringView, constructors taking in ASCIILiteral or Span are adjusted to
handle const lvalue references of those objects.

Different static integer values can now end up generating undesired symbols when
handled through references in calls to makeString() and makeAtomString(). This
can be avoided by explicitly copying the passed-in value.

* Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
(Inspector::InjectedScriptBase::makeCall):
(Inspector::InjectedScriptBase::makeAsyncCall):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::are8Bit):
(WTF::stringTypeAdapterAccumulator):
(WTF::tryMakeStringImplFromAdaptersInternal):
(WTF::handleWithAdapters):
(WTF::tryMakeStringFromAdapters):
(WTF::tryMakeString):
(WTF::makeString):
(WTF::tryMakeAtomStringFromAdapters):
(WTF::tryMakeAtomString):
(WTF::makeAtomString):
* Source/WTF/wtf/text/StringOperators.h:
(WTF::StringAppend::is8Bit const):
(WTF::StringAppend::writeTo const):
(WTF::StringAppend::length const):
(WTF::StringAppend::is8Bit): Deleted.
(WTF::StringAppend::writeTo): Deleted.
(WTF::StringAppend::length): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
</pre>

justinmichaud commented on Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
```diff
@@ -95,7 +95,7 @@ Ref<JSON::Value> InjectedScriptBase::makeCall(Deprecated::ScriptFunctionCall& fu
     auto resultJSONValue = toInspectorValue(globalObject, value);
     if (!resultJSONValue)
-        return JSON::Value::create(makeString("Object has too long reference chain (must not be longer than ", JSON::Value::maxDepth, ')'));
+        return JSON::Value::create(makeString("Object has too long reference chain (must not be longer than ", unsigned(JSON::Value::maxDepth), ')'));
```
I can never remember what the style guide says for this

zdobersek replied:
I can do `static_cast`.

fujii commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -463,13 +376,13 @@ RefPtr<StringImpl> tryMakeStringImplFromAdaptersInternal(unsigned length, bool a
 
 template<typename StringTypeAdapter, typename... StringTypeAdapters>
-String tryMakeStringFromAdapters(StringTypeAdapter adapter, StringTypeAdapters ...adapters)
+String tryMakeStringFromAdapters(const StringTypeAdapter& adapter, const StringTypeAdapters&... adapters)
```
We can remove the first argument `adapter`. But, this may be another task.

zdobersek committed:
Improve StringTypeAdapter templates and handling
https://bugs.webkit.org/show_bug.cgi?id=250017

Reviewed by Fujii Hironori.

StringConcatenate's makeString() and makeAtomString() are adjusted to accept
arguments as const lvalue references, propagating them down to each
StringTypeAdapter object without unnecessary copies. Similarly, the adapter
objects too are handled through const lvalue references once constructed and
passed on.

Newly-imposed constness requires additional const qualifiers on common methods
across different StringTypeAdapter specializations.

The are8Bit() and stringTypeAdapterAccumulator() helper functions are simplified
through using parameter packs and fold expressions to perform the necessary
operations across a range of adapter objects.

StringTypeAdapter specialization for StringView is moved from the StringView
header into the StringConcatenate header. A new partial specialization for the
StringTypeAdapter template is added, conditionally enabled if a const reference
of the specified type can be used to construct a StringView object. This
specialization inherits from the StringView specialization, hence a StringView
is constructed for such objects.

This enables removing a bunch of other StringTypeAdapter specializations.
Different WTF string and impl types can be used to construct the StringView,
as well as ASCIILiteral. For Vectors of character values, a corresponding Span
can be implicitly constructed and is then used for StringView construction.

In StringView, constructors taking in ASCIILiteral or Span are adjusted to
handle const lvalue references of those objects. A new templated StringView
constructor is added to handle C arrays of character data, delegating to the
corresponding character-and-length StringView constructors. The const char*
constructor is removed to avoid ambiguity with the constructor handling C arrays
of char values, with its now functionality inlined into the fromLatin1() helper.

Different static integer values can now end up generating undesired symbols when
handled through references in calls to makeString() and makeAtomString(). This
can be avoided by explicitly copying the passed-in value.

* Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
(Inspector::InjectedScriptBase::makeCall):
(Inspector::InjectedScriptBase::makeAsyncCall):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::are8Bit):
(WTF::stringTypeAdapterAccumulator):
(WTF::tryMakeStringImplFromAdaptersInternal):
(WTF::handleWithAdapters):
(WTF::tryMakeStringFromAdapters):
(WTF::tryMakeString):
(WTF::makeString):
(WTF::tryMakeAtomStringFromAdapters):
(WTF::tryMakeAtomString):
(WTF::makeAtomString):
* Source/WTF/wtf/text/StringOperators.h:
(WTF::StringAppend::is8Bit const):
(WTF::StringAppend::writeTo const):
(WTF::StringAppend::length const):
(WTF::StringAppend::is8Bit): Deleted.
(WTF::StringAppend::writeTo): Deleted.
(WTF::StringAppend::length): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::makeVersionedDirectoryPath):

Canonical link: https://commits.webkit.org/258946@main
